File: app/src/main/java/org/odddev/fantlab/core/validation/Validator.java
package org.odddev.fantlab.core.validation;

import android.databinding.BaseObservable;
import android.databinding.ObservableArrayMap;
import android.databinding.ObservableMap;

/**
 * @author kenrube
 * @since 26.11.16
 */

public abstract class Validator extends BaseObservable {

    public ObservableArrayMap<Integer, String> fields = new ObservableArrayMap<>();
    public ObservableArrayMap<Integer, String> fieldErrors = new ObservableArrayMap<>();

    protected Validator() {
        fields.addOnMapChangedCallback(
                new ObservableMap.OnMapChangedCallback<ObservableMap<Integer, String>, Integer, String>() {
                    @Override
                    public void onMapChanged(ObservableMap<Integer, String> integerStringObservableMap, Integer integer) {
                        validate(integer);
                    }
                });
    }

    protected abstract void validate(Integer field);

    protected abstract boolean areFieldsValid();
}

File: app/src/main/kotlin/ru/fantlab/android/core/validation/Validator.kt
package ru.fantlab.android.core.validation

import android.databinding.BaseObservable
import android.databinding.ObservableArrayMap
import android.databinding.ObservableMap

abstract class Validator protected constructor() : BaseObservable() {

	var fields = ObservableArrayMap<Int, String>()
	var fieldErrors = ObservableArrayMap<Int, String>()

	init {
		fields.addOnMapChangedCallback(
				object : ObservableMap.OnMapChangedCallback<ObservableMap<Int, String>, Int, String>() {
					override fun onMapChanged(integerStringObservableMap: ObservableMap<Int, String>, integer: Int) {
						validate(integer)
					}
				})
	}

	protected abstract fun validate(field: Int)

	protected abstract fun areFieldsValid(): Boolean
}